### PR TITLE
feat(java): implement codegen

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: src/server/supplements/recorder/java.ts [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+      case 'navigate':
+        return `navigate(${quote(action.url)})`;
+      case 'select':
+        return `selectOption(${quote(action.selector)}, ${formatSelectOption(action.options.length > 1 ? action.options : action.options[0])})`;
+    }
+  }
+
+  generateHeader(options: LanguageGeneratorOptions): string {
+    const formatter = new JavaScriptFormatter();
+    // BrowserContext context = browser.newContext(
```
Comment (yury-s) [inline src/server/supplements/recorder/java.ts:121]: Remove commented out code

File: src/server/supplements/recorder/java.ts [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+  const options: BrowserContextOptions = { ...device, ...contextOptions };
+  lines.push('new Browser.NewContextOptions()');
+  if (options.colorScheme)
+    lines.push(`  .withColorScheme(ColorScheme.${options.colorScheme.toUpperCase()})`);
+  if (options.geolocation)
+    lines.push(`  .withGeolocation(${options.geolocation.latitude}, ${options.geolocation.longitude})`);
+  if (options.locale)
+    lines.push(`  .withLocale("${options.locale}")`);
+  if (options.proxy)
+    lines.push(`  .withProxy(new Proxy("${options.proxy.server}"))`);
```
Comment (yury-s) [inline src/server/supplements/recorder/java.ts:186]:
```suggestion
    lines.push(`  .withProxy("${options.proxy.server}")`);
```

File: src/third_party/highlightjs/highlightjs/languages/java.js [hunk 3]
```diff
@@ -0,0 +1,176 @@
+// https://docs.oracle.com/javase/specs/jls/se15/html/jls-3.html#jls-3.10
```
Comment (yury-s) [inline src/third_party/highlightjs/highlightjs/languages/java.js:1]: Missing license header

File: test/cli/cli-codegen-2.spec.ts [hunk 4]
```diff
@@ -265,6 +294,14 @@ await Task.WhenAll(
   });
   await page.click('text=click me');`);
 
+    expect(sources.get('<java>').text).toContain(`
+      // Click text=click me
+      page.onceDialog(dialog -> {
```
Comment (yury-s) [inline test/cli/cli-codegen-2.spec.ts:299]: there are no `once*` methods in java

**Review by yury-s — APPROVED** [review]

File: src/server/supplements/recorder/java.ts [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+import { JavaScriptFormatter } from './javascript';
+
+export class JavaLanguageGenerator implements LanguageGenerator {
+  id = 'java';
+  fileName = '<java>';
+  highlighter = 'java';
+
+  generateAction(actionInContext: ActionInContext): string {
+    const { action, pageAlias } = actionInContext;
+    const formatter = new JavaScriptFormatter(6);
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply src/server/supplements/recorder/java.ts:32]: This is initial indent, inside 3 nested blocks...

File: src/server/supplements/recorder/java.ts (hunk 2 above)
Comment (pavelfeldman) [reply src/server/supplements/recorder/java.ts:186]: Leaving as agreed offline

File: src/third_party/highlightjs/highlightjs/languages/java.js (hunk 3 above)
Comment (pavelfeldman) [reply src/third_party/highlightjs/highlightjs/languages/java.js:1]: This is generated by the highlighter.js build, license is below...

File: test/cli/cli-codegen-2.spec.ts (hunk 4 above)
Comment (pavelfeldman) [reply test/cli/cli-codegen-2.spec.ts:299]: There should be this one :)

**Pull request merged**